### PR TITLE
[SPARK-35719][SQL] Support type conversion between timestamp and timestamp without time zone type

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

1. Extend the Cast expression and support TimestampType in casting to TimestampWithoutTZType.
2. There was a mistake in casting TimestampWithoutTZType as TimestampType in https://github.com/apache/spark/pull/32864. The target value should be `sourceValue - timeZoneOffset` instead of being the same value.

### Why are the changes needed?

To conform the ANSI SQL standard which requires to support such casting.

### Does this PR introduce _any_ user-facing change?

No, the new timestamp type is not released yet.

### How was this patch tested?

Unit test

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 1]
```diff
@@ -513,6 +514,8 @@ abstract class CastBase extends UnaryExpression with TimeZoneAwareExpression wit
   private[this] def castToTimestampWithoutTZ(from: DataType): Any => Any = from match {
     case DateType =>
       buildCast[Int](_, d => daysToMicros(d, ZoneOffset.UTC))
+    case TimestampType =>
+      buildCast[Long](_, ts => ts)
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:518]: Should we consider the input timestamp w/o tz as local offset. How about the chain:
input micros -> instant (UTC) -> local timestamp (UTC) = local timestamp (session TZ) -> instant -> result in micros

In your implementation, you say that timestamp w/o TZ is a timestamp in UTC, and it is not local timestamp.

Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:518]: You are right. I will fix it soon. Thanks!

**Comment by gengliangwang** [comment]
I will fix `timestamp without time zone` => `timestamp` in this PR, too.

**Comment by gengliangwang** [comment]
@MaxGekk @cloud-fan this is ready for review again. The conversions is a bit tricky. PTAL.

**Review by MaxGekk — APPROVED** [review]
+1, LGTM. Merging to master.
Thank you, @gengliangwang .

**Pull request closed without merging**